# 집계값 features로 overSampling

id 별로 groupby하여 집계값을 지정할 경우 데이터의 개수가 3125개로 너무 적음.   
하지만 제출할 submission의 index값이 id로 지정되어 test 역시 grouping 이 불가피한 것으로 생각됨.  
train set을 더 확장함으로서 과적합을 방지하고 모델의 고도화를 기대함.   

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

import time
import datetime
pd.set_option('display.max_columns', None)

train = pd.read_csv('train_features.csv')
test = pd.read_csv('test_features.csv')
train_label = pd.read_csv('train_labels.csv')
submission = pd.read_csv('sample_submission.csv')

In [7]:
train_label.label_desc.value_counts()

Non-Exercise                                    1518
Device on Table                                   97
Plank                                             55
Wall Squat                                        48
Jump Rope                                         47
                                                ... 
Triceps Kickback (knee on bench) (right arm)      13
Arm Band Adjustment                               12
Dumbbell Row (knee on bench) (left arm)           12
Triceps Kickback (knee on bench) (left arm)       12
Dumbbell Row (knee on bench) (right arm)          12
Name: label_desc, Length: 61, dtype: int64

In [9]:
train_label

id  label                          label_desc
0        0     37           Shoulder Press (dumbbell)
1        1     26                        Non-Exercise
2        2      3                  Biceps Curl (band)
3        3     26                        Non-Exercise
4        4     26                        Non-Exercise
...    ...    ...                                 ...
3120  3120     26                        Non-Exercise
3121  3121     26                        Non-Exercise
3122  3122     15  Dynamic Stretch (at your own pace)
3123  3123     26                        Non-Exercise
3124  3124      2                          Bicep Curl

[3125 rows x 3 columns]

In [10]:
y = train_label['label']

In [24]:
train_200 = pd.read_csv('train_aggs_200Times.csv', index_col = 'id')
test_200 = pd.read_csv('test_aggs_200Times.csv',index_col = 'id')

In [23]:
train_200

acc_x_amax  acc_x_amin  acc_x_mean  acc_x_var  acc_x_median  \
id                                                                  
0       1.343113    0.613582    0.939629   0.040701      0.980503   
1      -0.059166   -2.156208   -0.541579   0.256409     -0.256984   
2       1.219836   -1.142847    0.206800   0.516682      0.579793   
3      -0.679208   -1.339996   -0.918718   0.013523     -0.919209   
4       0.599720   -0.831264   -0.079400   0.091618     -0.076026   
...          ...         ...         ...        ...           ...   
3120    0.390798   -1.624711   -0.288550   0.247556     -0.106967   
3121   -0.446650   -1.575455   -0.981668   0.058628     -0.974358   
3122    0.744666   -2.141936   -1.095197   0.337772     -1.077250   
3123    0.164426   -0.929133   -0.316117   0.071350     -0.290501   
3124    0.538809   -0.980186   -0.275555   0.295196     -0.282074   

      acc_x_percentile_25  acc_x_percentile_75  acc_y_amax  acc_y_amin  \
id                                                                       
0                0.736127             1.090305    0.133165   -0.624113   
1               -0.877781            -0.150953    0.495337   -0.997057   
2               -0.464966             0.815432    0.469709   -0.690528   
3               -0.986731            -0.851360    0.210985   -0.540827   
4               -0.354173             0.157189    0.220760   -2.055076   
...                   ...                  ...         ...         ...   
3120            -0.779654             0.117437   -0.027536   -1.289257   
3121            -1.159156            -0.803324    0.117965   -0.600138   
3122            -1.558972            -0.701768    1.268138   -2.036646   
3123            -0.538415            -0.114130    1.471337    0.397902   
3124            -0.841061             0.266734   -0.305105   -0.891517   

      acc_y_mean  acc_y_var  acc_y_median  acc_y_percentile_25  \
id                                                               
0      -0.240798   0.034805     -0.246550            -0.380588   
1      -0.413170   0.149589     -0.584535            -0.741603   
2      -0.102568   0.026990     -0.082992            -0.187202   
3      -0.173210   0.042281     -0.199728            -0.318023   
4      -0.905563   0.105969     -0.921701            -1.004630   
...          ...        ...           ...                  ...   
3120   -0.681382   0.055939     -0.698590            -0.819203   
3121   -0.192524   0.017643     -0.176790            -0.279944   
3122   -0.414260   0.136888     -0.333738            -0.575998   
3123    0.885078   0.031748      0.876825             0.774396   
3124   -0.667109   0.031278     -0.740498            -0.816293   

      acc_y_percentile_75  acc_z_amax  acc_z_amin  acc_z_mean  acc_z_var  \
id                                                                         
0               -0.085563   -0.054876   -0.672738   -0.344088   0.021236   
1               -0.077095    0.861091   -0.522313    0.313459   0.159479   
2               -0.022000    1.088743    0.073846    0.558121   0.069127   
3                0.061739    0.410242   -0.181537    0.148455   0.013527   
4               -0.817032    1.057852   -1.231836    0.186689   0.108823   
...                   ...         ...         ...         ...        ...   
3120            -0.537470    0.849966   -0.551185    0.194684   0.175208   
3121            -0.093278    0.176047   -0.403201   -0.107707   0.009315   
3122            -0.182982    0.803950   -0.459267    0.278428   0.049012   
3123             0.993590    0.382916   -0.646863   -0.132612   0.090447   
3124            -0.528173    0.543977   -0.163281    0.312278   0.042212   

      acc_z_median  acc_z_percentile_25  acc_z_percentile_75   gy_x_amax  \
id                                                                         
0        -0.316034            -0.413650            -0.236843   30.962266   
1         0.448468            -0.060086             0.597465  176.742

`train_200`과 `test_200` 데이터셋은 aggFeatures2.ipynb에서 생성한 데이터셋을 csv로 저장하여 불러온 것이다.
id별로 `[np.max, np.min, np.mean, np.var, np.median, percentile(25), percentile(75)]`을 집계 계산한 결과이다.  

---

train_label을 확인해보면 총 label의 종류는 61가지, 그리고 그 중 운동이 아닌 ('Non-Exercise')가 가장 많다.  
그 다음으로 많은 'Device on Table'의 개수는 97개에 비해 Non-Exercise는 1518개로 레이블이 불균형하다는 것을 알 수 있다.    

--- 
이상 레이블을 검출하는 문제라고 생각했을 때, 이상 레이블을 가진 데이터의 수가 너무 적어 모델이 정상 레이블에 치우쳐진 학습이 이루어질 수 있다. 그로 인해 제대로된 이상 데이터 검출이 어려워지기 쉽다. 이를 해결하기 위해 Under Sampling 또는 Over Sampling 방법을 활용할 수 있다. Over Sampling 방식은 이상 데이터와 같이 적은 데이터 세트를 증식하여 학습을 위한 충분한 데이터를 확보하는 방법이다. `SMOTE`는 적은 데이터 세트에 있는 개별 데이터들의 K 최근접 이웃(K Nearest Neighbor)을 찾아 이 데이터와 K개의 이웃들의 차이를 일정 값으로 만들어서 기존 데이터와 약간 차이가 나는 새로운 데이터들을 생성하는 방식이다. 

---

Non-Exercise의 개수가 1518개고 나머지 labels들을 모두 1518개만큼 추가로 데이터를 생성해는 것이다. 그래서 총 SMOTE로 `fit_sample()`로 생성된 총 데이터 개수는 1518 * 61 = 92,598개. Over Sampling 하기 전 학습에 사용된 데이터의 크기는 3125개로 적은 개수였다. 그러나 Over Sampling 로 추가 생성되어 학습에 필요한 데이터을 가지고 있음으로 그동안 사용했던 XGBoost 모델과 더불어 LightGBM 모델을 사용해도 가능할 것으로 보인다.  

In [26]:
from imblearn.over_sampling import SMOTE

In [27]:
# SMOTE oversampling

smote = SMOTE(random_state = 0 )

In [28]:
X_train_over, y_train_over = smote.fit_sample(train_200, y)

In [29]:
print('SMOTE 적용 전 학습용 feature/label 데이터 세트 : ', train_200.shape, y.shape)
print('SMOTE 적용 후 학습용 feature/label 데이터 세트 : ', X_train_over.shape, y_train_over.shape)

SMOTE 적용 전 학습용 feature/label 데이터 세트 :  (3125, 126) (3125,)
SMOTE 적용 후 학습용 feature/label 데이터 세트 :  (92598, 126) (92598,)


In [30]:
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split

In [31]:
X_train, X_test, y_train, y_test = train_test_split(X_train_over, y_train_over, test_size = 0.2, 
                                                   random_state = 42)
print(y_train.nunique(), y_test.nunique())

61 61


In [33]:
xgb = XGBClassifier(n_estimators = 400, learning_rate = 0.05, 
                    max_depth = 6, 
                    objective = 'multi:softprob', random_state = 0) 

In [37]:
evals = [(X_test.values, y_test)]

In [39]:
X_test

acc_x_amax  acc_x_amin  acc_x_mean  acc_x_var  acc_x_median  \
20899   -0.475059   -1.580237   -0.997151   0.106090     -1.057743   
65414    0.075172   -1.046176   -0.453056   0.072646     -0.415666   
50007   -0.313768   -0.867212   -0.604431   0.014134     -0.591677   
85634    1.331899   -0.545366    0.305835   0.166266      0.396090   
63102    1.538876   -1.792806   -0.063857   0.293877     -0.045943   
...           ...         ...         ...        ...           ...   
15513   -0.105903   -0.118900   -0.112220   0.000005     -0.112243   
53102    0.785924   -3.321788   -0.902767   1.054088     -0.812913   
3295     0.635939   -0.522834    0.104473   0.090619      0.063070   
47844   -0.770254   -1.133933   -0.953068   0.004274     -0.959130   
90388    6.354024   -6.446741    0.783625   1.603389      1.045949   

       acc_x_percentile_25  acc_x_percentile_75  acc_y_amax  acc_y_amin  \
20899            -1.276677            -0.725587    0.509108   -0.484729   
65414            -0.630809            -0.252494   -0.258833   -1.475329   
50007            -0.615211            -0.565375    0.411527   -0.606079   
85634            -0.051842             0.624604    2.267406   -1.359774   
63102            -0.406367             0.245478    1.738472   -1.901600   
...                    ...                  ...         ...         ...   
15513            -0.113912            -0.110589    0.008552   -0.009222   
53102            -1.556625            -0.083405    1.028507   -1.958342   
3295             -0.088477             0.260257    0.011802   -1.425805   
47844            -0.990209            -0.911939    0.300797   -0.552815   
90388             0.688236             1.286545    1.784489   -1.807549   

       acc_y_mean  acc_y_var  acc_y_median  acc_y_percentile_25  \
20899    0.002880   0.036382      0.015780            -0.139783   
65414   -0.831582   0.108327     -0.819853            -1.051015   
50007   -0.318216   0.089047     -0.404306            -0.447674   
85634    0.014823   1.140297     -0.475525            -0.897666   
63102   -0.390941   0.758305     -0.730143            -0.950575   
...           ...        ...           ...                  ...   
15513    0.000240   0.000007      0.000402            -0.001305   
53102   -0.469020   0.472702     -0.403798            -1.021284   
3295    -0.770761   0.071629     -0.807219            -0.923766   
47844   -0.090241   0.024515     -0.083511            -0.172290   
90388   -0.613443   0.399301     -0.798133            -1.078667   

       acc_y_percentile_75  acc_z_amax  acc_z_amin  acc_z_mean  acc_z_var  \
20899             0.142720    0.321403   -1.210364   -0.135338   0.080662   
65414            -0.524855    0.691279   -0.218095    0.142839   0.030677   
50007            -0.251899   -0.067779   -0.669766   -0.284631   0.020236   
85634             1.084355    0.261175   -0.981635   -0.192713   0.035458   
63102             0.099569    0.718813   -1.212479    0.002637   0.210242   
...                    ...         ...         ...         ...        ...   
15513             0.001481    0.874501    0.852165    0.865514   0.000015   
53102             0.066529    2.116917   -0.724653    0.306289   0.416363   
3295             -0.595918    0.540267   -0.935533   -0.306991   0.148536   
47844             0.010933    0.299324   -0.796538   -0.079480   0.066570   
90388            -0.275084    1.660659   -1.037652    0.133771   0.250653   

       acc_z_median  acc_z_percentile_25  acc_z_percentile_75   gy_x_amax  \
20899     -0.095962            -0.318168             0.081474   93.896921   
65414      0.125889             0.018855             0.251726   87.455445   
50007     -0.268738            -0.337056            -0.231551  139.944590   
85634     -0.172095            -0.309494            -0.068726  154.602973   
63102      0.182902            -0.361741             0.307764  285.113484   
...             ...                  ...                  ...         ...

In [40]:
start = time.time()
xgb.fit(X_train_over.values, y_train_over,
            eval_metric = 'mlogloss', 
            eval_set= evals, 
            early_stopping_rounds = 100)

sec = time.time()-start
times = str(datetime.timedelta(seconds = sec)).split('.')

print('소요시간 : ', times)

[0]	validation_0-mlogloss:2.90889
[1]	validation_0-mlogloss:2.53030
[2]	validation_0-mlogloss:2.26883
[3]	validation_0-mlogloss:2.06737
[4]	validation_0-mlogloss:1.90158
[5]	validation_0-mlogloss:1.76174
[6]	validation_0-mlogloss:1.63911
[7]	validation_0-mlogloss:1.53177
[8]	validation_0-mlogloss:1.43617
[9]	validation_0-mlogloss:1.34949
[10]	validation_0-mlogloss:1.27109
[11]	validation_0-mlogloss:1.19940
[12]	validation_0-mlogloss:1.13359
[13]	validation_0-mlogloss:1.07303
[14]	validation_0-mlogloss:1.01684
[15]	validation_0-mlogloss:0.96458
[16]	validation_0-mlogloss:0.91582
[17]	validation_0-mlogloss:0.87004
[18]	validation_0-mlogloss:0.82739
[19]	validation_0-mlogloss:0.78721
[20]	validation_0-mlogloss:0.74932
[21]	validation_0-mlogloss:0.71379
[22]	validation_0-mlogloss:0.68004
[23]	validation_0-mlogloss:0.64814
[24]	validation_0-mlogloss:0.61802
[25]	validation_0-mlogloss:0.58962
[26]	validation_0-mlogloss:0.56249
[27]	validation_0-mlogloss:0.53680
[28]	validation_0-mlogloss:0.5

In [41]:
y_pred = xgb.predict_proba(test_200.values)
submission.iloc[:,1:] = y_pred
submission.to_csv('submission파일/over_xgb200T.csv', index =False)

제출 결과값 : 0.834271837   
Over Sampling을 하지 않은 데이터셋으로 학습한 모델이 더 좋은 결과값을 가짐. 

In [42]:
# LightGBM
from lightgbm import LGBMClassifier
lgbm = LGBMClassifier(n_estimators = 400, learning_rate = 0.05, random_state = 0)
start = time.time()
lgbm.fit(X_train_over, y_train_over, eval_set = evals,
        early_stopping_rounds = 100)
sec = time.time() -start 

print('수행시간 : ', str(datetime.timedelta(seconds = sec)).split('.'))

[1]	valid_0's multi_logloss: 1.57191
Training until validation scores don't improve for 100 rounds
[2]	valid_0's multi_logloss: 1.34233
[3]	valid_0's multi_logloss: 1.17811
[4]	valid_0's multi_logloss: 1.04513
[5]	valid_0's multi_logloss: 0.931983
[6]	valid_0's multi_logloss: 0.835131
[7]	valid_0's multi_logloss: 0.750146
[8]	valid_0's multi_logloss: 0.676082
[9]	valid_0's multi_logloss: 0.610327
[10]	valid_0's multi_logloss: 0.551632
[11]	valid_0's multi_logloss: 0.499288
[12]	valid_0's multi_logloss: 0.452491
[13]	valid_0's multi_logloss: 0.410426
[14]	valid_0's multi_logloss: 0.372459
[15]	valid_0's multi_logloss: 0.33831
[16]	valid_0's multi_logloss: 0.307418
[17]	valid_0's multi_logloss: 0.279412
[18]	valid_0's multi_logloss: 0.254094
[19]	valid_0's multi_logloss: 0.231141
[20]	valid_0's multi_logloss: 0.210416
[21]	valid_0's multi_logloss: 0.191407
[22]	valid_0's multi_logloss: 0.174318
[23]	valid_0's multi_logloss: 0.158783
[24]	valid_0's multi_logloss: 0.144657
[25]	valid_0's m

In [43]:
y_pred = lgbm.predict_proba(test_200.values)
submission.iloc[:,1:] = y_pred
submission.to_csv('submission파일/over_lgbm200T.csv', index = False)

제출결과값 : 1.1567946727

In [45]:
def percentile(n):
    def percentile_(x):
        return np.percentile(x, n)
    percentile_.__name__ = 'percentile_%s' % n
    return percentile_

In [47]:
id_train = train.drop('time',axis = 1).groupby('id').agg([np.max, np.min, np.mean, 
                       np.var, np.median,
                      percentile(25), percentile(75)])

In [49]:
smote = SMOTE(random_state = 0 )
X_over, y_over = smote.fit_sample(id_train, y)

In [50]:
X_over.shape

(92598, 42)

In [51]:
X_train, X_test, y_train, y_test = train_test_split(X_over, y_over, test_size = 0.2,
                                                    random_state = 42)
evals = [(X_test.values, y_test)]

In [54]:
start = time.time()
xgb.fit(X_over.values, y_over, 
       eval_metric = 'mlogloss',
       eval_set = evals,
       early_stopping_rounds = 100)
sec = time.time()-start
print('소요시간 : ', str(datetime.timedelta(seconds = sec)).split('.'))

[0]	validation_0-mlogloss:2.94802
[1]	validation_0-mlogloss:2.56972
[2]	validation_0-mlogloss:2.30743
[3]	validation_0-mlogloss:2.10514
[4]	validation_0-mlogloss:1.93954
[5]	validation_0-mlogloss:1.79955
[6]	validation_0-mlogloss:1.67840
[7]	validation_0-mlogloss:1.57161
[8]	validation_0-mlogloss:1.47642
[9]	validation_0-mlogloss:1.39064
[10]	validation_0-mlogloss:1.31301
[11]	validation_0-mlogloss:1.24189
[12]	validation_0-mlogloss:1.17677
[13]	validation_0-mlogloss:1.11616
[14]	validation_0-mlogloss:1.05996
[15]	validation_0-mlogloss:1.00773
[16]	validation_0-mlogloss:0.95941
[17]	validation_0-mlogloss:0.91392
[18]	validation_0-mlogloss:0.87130
[19]	validation_0-mlogloss:0.83125
[20]	validation_0-mlogloss:0.79312
[21]	validation_0-mlogloss:0.75742
[22]	validation_0-mlogloss:0.72374
[23]	validation_0-mlogloss:0.69153
[24]	validation_0-mlogloss:0.66132
[25]	validation_0-mlogloss:0.63251
[26]	validation_0-mlogloss:0.60513
[27]	validation_0-mlogloss:0.57919
[28]	validation_0-mlogloss:0.5

In [55]:
id_test = test.drop('time',axis = 1).groupby('id').agg([np.max, np.min, np.mean, 
                       np.var, np.median,
                      percentile(25), percentile(75)])

In [56]:
y_pred = xgb.predict_proba(id_test.values)
submission.iloc[:,1:]=y_pred
submission.to_csv('submission파일/over_xgb.csv',index =False)

제출 결과값 : 1.0243872569